src/dbnode/storage/index/compaction/compactor.go

```diff
@@ -0,0 +1,349 @@
+		}
+		if err != nil {
+			return err
+		}
+
+		// Reset docs batch for reuse
+		batch = batch[:0]
+		return nil
+	}
+
+	for _, seg := range segs {
+		reader, err := seg.Reader()
+		if err != nil {
+			return nil, err
+		}
+
+		iter, err := reader.AllDocs()
+		if err != nil {
+			return nil, err
+		}
+
+		for iter.Next() {
+			batch = append(batch, iter.Current())
+			if len(batch) < c.docsMaxBatch {
+				continue
+			}
+			if err := flushBatch(); err != nil {
+				return nil, err
+			}
+		}
+
+		if err := iter.Err(); err != nil {
+			return nil, err
+		}
+		if err := iter.Close(); err != nil {
+			return nil, err
+		}
+		if err := reader.Close(); err != nil {
+			return nil, err
+		}
+	}
+
+	// Last flush in case some remaining docs that
+	// weren't written to the mutable segment
+	if err := flushBatch(); err != nil {
+		return nil, err
+	}
+
+	return c.compactFromBuilderWithLock(builder)
+}
+
+func (c *Compactor) compactFromBuilderWithLock(
+	builder segment.Builder,
+) (segment.Segment, error) {
+	defer func() {
+		// Release resources regardless of result,
+		// otherwise old compacted segments are held onto
+		// strongly
+		builder.Reset(0)
+	}()
+
+	// Since this builder is likely reused between compaction
+	// runs, we need to copy the docs slice
+	allDocs := builder.Docs()
+	if len(allDocs) == 0 {
+		return nil, errCompactorBuilderEmpty
+	}
+
+	allDocsCopy := make([]doc.Document, len(allDocs))
+	copy(allDocsCopy, allDocs)
+
+	err := c.writer.Reset(builder)
+	if err != nil {
+		return nil, err
+	}
+
+	success := false
+	closers := new(closers)
+	fstData := fst.SegmentData{
+		MajorVersion: c.writer.MajorVersion(),
+		MinorVersion: c.writer.MinorVersion(),
+		Metadata:     append([]byte(nil), c.writer.Metadata()...),
+		DocsReader:   docs.NewSliceReader(0, allDocsCopy),
+		Closer:       closers,
+	}
+
+	// Cleanup incase we run into issues
+	defer func() {
+		if !success {
+			closers.Close()
+		}
+	}()
+
+	c.buff.Reset()
+	if err := c.writer.WritePostingsOffsets(c.buff); err != nil {
+		return nil, err
+	}
+
+	fstData.PostingsData, err = c.mmapAndAppendCloser(c.buff.Bytes(), closers)
+	if err != nil {
+		return nil, err
+	}
+
+	c.buff.Reset()
+	if err := c.writer.WriteFSTTerms(c.buff); err != nil {
+		return nil, err
+	}
+
+	fstData.FSTTermsData, err = c.mmapAndAppendCloser(c.buff.Bytes(), closers)
+	if err != nil {
+		return nil, err
+	}
+
+	c.buff.Reset()
+	if err := c.writer.WriteFSTFields(c.buff); err != nil {
+		return nil, err
+	}
+
+	fstData.FSTFieldsData, err = c.mmapAndAppendCloser(c.buff.Bytes(), closers)
+	if err != nil {
+		return nil, err
+	}
+
+	compacted, err := fst.NewSegment(fstData, c.fstOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	success = true
+
+	return compacted, nil
+}
+
+func (c *Compactor) mmapAndAppendCloser(
+	fromBytes []byte,
+	closers *closers,
+) ([]byte, error) {
+	// Copy bytes to new mmap region to hide from the GC
+	mmapedResult, err := mmap.Bytes(int64(len(fromBytes)), mmap.Options{
```

Comment:
will all these mmaps usually at least a few KiB in size? Every mmap costs us at least one full page so we should be careful to avoid allocating lots of them for small byte slices

Reply:
Yeah most of these will be like 10mb at least, it's 3 really huge files basically.